### PR TITLE
fix(#2056): date picker reset issues (React fix)

### PR DESCRIPTION
**Pull request by syedszeeshan**
# Before (the change)

# After (the change)
The value is resetting properly.

## Make sure that you've checked the boxes below before you submit the PR

- [x] I have read and followed the [setup steps](https://goa-dio.atlassian.net/wiki/spaces/DS/pages/3086385489/Contribution+-+web+content#%5BinlineExtension%5DCode-contributions)
- [ ] I have created necessary unit tests
- [x] I have tested the functionality in both React and Angular.

## Steps needed to test

**Angular Code**

```
import { Component } from "@angular/core";
import { FormControl, FormGroup, Validators } from "@angular/forms";

@Component({
  selector: "abgov-datepicker",
  templateUrl: "./datepicker..component.html",
  styleUrls: ["./datepicker.component.css"],
})
export class DatePickerComponent {
  dateCtrl = new FormControl();

  setValue() {
    const d = new Date();
    d.setDate(d.getDate() - 7);
    this.dateCtrl.patchValue(d);
    console.log("setValue called: ", this.dateCtrl.value);
  }

  clearValue() {
    this.dateCtrl.reset();
    console.log("reset called: ", this.dateCtrl.value);
  }
}

------------------------------

<goa-form-item label="Input Label" helptext="Helper text" mb="3xl">
  <goa-date-picker [formControl]="dateCtrl" [value]="dateCtrl.value" ngDefaultControl />
</goa-form-item>
<br />
<goa-button (_click)="setValue()" mr="l">Set Value</goa-button>
<goa-button (_click)="clearValue()">Clear Value</goa-button>

```

**React Code**
```
import { GoADatePicker, GoAFormItem, GoAButton } from "@abgov/react-components";
import { useState } from "react";

export function Datepicker() {
  const [date, setDate] = useState<Date | undefined>();

  const setNewDate = (value: Date) => {
    setDate(value);
  };

  function setValue() {
    const d = new Date();
    d.setDate(d.getDate() - 7);

    setDate(d);
  }

  function clearValue() {
    setDate(undefined);
  }

  return (
    <>
      <GoAFormItem label="Item">
        <GoADatePicker
          name="item"
          value={date}
          onChange={(_, value) => setNewDate(value)}
          mb="xl"
        ></GoADatePicker>
      </GoAFormItem>
      <br />
      <GoAButton onClick={setValue} mr="l">
        Set Value
      </GoAButton>
      <GoAButton onClick={clearValue}>Clear Value</GoAButton>
    </>
  );
}
```

### CODE REVIEW

**Comment by syedszeeshan** [comment]
### Fallback Plan (Code for 'goaDateValue' custom directive)

**value-directive.ts**
```
import {
  forwardRef,
  Directive,
  ElementRef,
  HostListener,
  Renderer2,
} from "@angular/core";
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from "@angular/forms";

@Directive({
  selector: "[goaValue]",
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => ValueDirective),
      multi: true,
    },
  ],
})
export class ValueDirective implements ControlValueAccessor {
  private _value = "";

  /* eslint-disable @typescript-eslint/no-explicit-any */
  onChange: any = () => {};
  onTouched: any = () => {};

  get value(): string {
    return this._value;
  }

  set value(val: string) {
    this._value = val;
    this.onChange(this._value);
    this.onTouched();
    this.elementRef.nativeElement.value = val;
  }

  writeValue(value: string) {
    this.value = value;
  }

  registerOnChange(fn: () => void) {
    this.onChange = fn;
  }

  registerOnTouched(fn: () => void) {
    this.onTouched = fn;
  }

  constructor(protected elementRef: ElementRef) {}

  @HostListener("_change", ["$event.detail.value"])
  listenForValueChange(value: string) {
    this.value = value;
  }
}

@Directive({
  selector: "[goaValueList]",
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => ValueListDirective),
      multi: true,
    },
  ],
})
export class ValueListDirective implements ControlValueAccessor {
  private _value?: string[] = [];

  onChange: any = () => {};
  onTouched: any = () => {};

  get value(): string[] | undefined {
    return this._value;
  }

  set value(val: string[] | undefined) {
    if (val && val !== this._value) {
      this._setValue(val);
      this.elementRef.nativeElement.value = JSON.stringify(val);
    }
  }

  writeValue(value?: string[]) {
    if (value) {
      this.value = value;
    }
  }

  registerOnChange(fn: () => void) {
    this.onChange = fn;
  }

  registerOnTouched(fn: () => void) {
    this.onTouched = fn;
  }

  constructor(protected elementRef: ElementRef) {}

  @HostListener("_change", ["$event.detail.value"])
  listenForValueChange(value: string) {
    if (!value) {
      this._setValue(undefined);
      return;
    }

    try {
      this.value = JSON.parse(value);
    } catch (e) {
      // we still need to trigger the events to prevent any previous valid value to remain set.
      const v = value.match(/^[\w\s,]*$/) ? value.split(",") : undefined;
      this._setValue(v);
    }
  }

  _setValue(value?: string[]) {
    this._value = value;
    this.onChange(value);
    this.onTouched();
  }
}

@Directive({
  selector: "[goaDateValue]",
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => DateValueDirective),
      multi: true,
    },
  ],
})
export class DateValueDirective implements ControlValueAccessor {
  private _value: string | null = null;

  onChange: any = () => {};
  onTouched: any = () => {};

  get value(): string | null {
    return this._value;
  }

  set value(val: string | null) {
    if (this._value !== val) {
      this._value = val;
      this.onChange(this._value);
      this.onTouched();
      this.updateElementValue(val);
    }
  }

  writeValue(value: string | null) {
    if (this._value !== value) {
      this._value = value;
      this.updateElementValue(value);
    }
  }

  registerOnChange(fn: () => void) {
    this.onChange = fn;
  }

  registerOnTouched(fn: () => void) {
    this.onTouched = fn;
  }

  constructor(
    protected elementRef: ElementRef,
    private renderer: Renderer2,
  ) {}

  @HostListener("_change", ["$event.detail.value"])
  listenForValueChange(value: string | null) {
    this.value = value;
  }

  private updateElementValue(value: string | null) {
    this.renderer.setProperty(this.elementRef.nativeElement, "value", value || "");
  }
}

```

**angular-components.module.ts**
```
import { CUSTOM_ELEMENTS_SCHEMA, NgModule } from "@angular/core";
import { ValueDirective, ValueListDirective, DateValueDirective } from "./value-directive";
import { CheckedDirective } from "./checked-directive";

@NgModule({
  declarations: [ValueDirective, ValueListDirective, DateValueDirective, CheckedDirective],
  exports: [ValueDirective, ValueListDirective, DateValueDirective, CheckedDirective],
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
})
export class AngularComponentsModule {}

export { ValueDirective, ValueListDirective, DateValueDirective, CheckedDirective };
```

**Comment by lizhuomeng71** [comment]
I notice something new on angular, I am still working on this ticket, moving this ticket back to in-testing

**Comment by lizhuomeng71** [comment]
angular issue is resolved, I missed something on my html, now is good to review again

File: libs/react-components/src/lib/date-picker/date-picker.tsx [hunk 1]
```diff
@@ -51,21 +51,22 @@ export function GoADatePicker({
...
-    }
+    };
   }, [onChange]);
 
   return (
     <goa-date-picker
       ref={ref}
       name={name}
-      value={value?.toISOString()}
+      value={value ? value.toISOString() : ""}
```
Comment (chrisolsen) [inline libs/react-components/src/lib/date-picker/date-picker.tsx:69]: Does `value?.toISOString() || ""` not work?

Comment (syedszeeshan) [reply libs/react-components/src/lib/date-picker/date-picker.tsx:69]: Correct, that does not work.

Comment (syedszeeshan) [reply libs/react-components/src/lib/date-picker/date-picker.tsx:69]: https://github.com/user-attachments/assets/952da346-b00b-485c-83e2-3044bd76f2d3

Comment (syedszeeshan) [reply libs/react-components/src/lib/date-picker/date-picker.tsx:72]: Correct, this `value={value?.toISOString()}` does not work. (video below)
https://github.com/user-attachments/assets/d2517284-50c5-4f1f-9dd6-7c1730217f6b

**Review by chrisolsen — APPROVED** [review]

**Review by chrisolsen — APPROVED** [review]

**Pull request merged**